lib/dao.js

```diff
@@ -509,6 +516,35 @@ DataAccessObject.upsert = function(data, options, cb) {
   if (id === undefined || id === null) {
     return this.create(data, options, cb);
   }
+  var doValidate = undefined;
+  if (options.validate === undefined) {
+    if (Model.settings.validateUpsert === undefined) {
+      if (Model.settings.automaticValidation !== undefined) {
+        doValidate = Model.settings.automaticValidation;
+      }
+    } else {
+      doValidate = Model.settings.validateUpsert;
+    }
+  } else {
+    doValidate = options.validate;
+  }
+
+  var forceId = this.settings.forceId;
+  if (forceId) {
+    options = Object.create(options);
+    options.validate = !!doValidate;
+    if (doValidate) {
+      Model.findById(id, options, function(err, model) {
+        if (err) return cb(err);
+        if (!model) return cb(errorModelNotFound(id));
+        model.updateAttributes(data, options, cb);
+      });
+    } else {
+      const model = new Model({id: id}, {persisted: true});
+      model.updateAttributes(data, options, cb);
+    }
+    return;
```

Comment:
I think this will break promise mode, you need to `return cb.promise`. @loay could you please check?

A fix should include a unit-test to verify.